### ISSUE TITLE
TwoStep Learning

### ISSUE THREAD

**Issue by xixihahaggg** [mpariente/asteroid#366]
Nice to try this wonderful project!

I have some questions about implementation of the TwoStep learning.

1. The first stage is training the encoder and decoder, and since the mask is directly derived by the softmax function, I'm not sure why use the PIT in the loss function, I think the order is fix.

2. I tried to use the make_enc_dec to build the encoder and decoder instead of the default setting in the recipe. But I found performance is converge at around 8 dB si-sdr, which was much worse than the Adapitive1Dconv and deconv in the model.py. I'm not sure what's going here here.

Very appreciate your answer!

**Comment by mpariente** [comment]
1. You're probably right. 
2. Can you link the code you used?

**Comment by xixihahaggg** [comment]
Thank you for your response!
```
class EncoderDecoder(nn.Module):
    def __init__(self, fb_name="free",
                 kernel_size=16,
                 n_filters=512,
                 stride=8,
                 encoder_activation=None,
                 sample_rate=8000, n_sources=2,
                 **fb_kwargs):
        super().__init__()

        self.encoder, self.decoder = make_enc_dec(
            fb_name,
            kernel_size=kernel_size,
            n_filters=n_filters,
            stride=stride,
            sample_rate=sample_rate,
            **fb_kwargs,
        )

    def forward(self, mixture, clean_sources):
        enc_mixture = F.relu(self.encoder(mixture.unsqueeze(1)))
        enc_masks = F.softmax(self.encoder(clean_sources), dim=1)
        s_recon_enc = enc_masks * enc_mixture.unsqueeze(1)
        recon_sources = self.decoder(s_recon_enc)
        return pad_x_to_y(recon_sources, clean_sources), enc_masks
```
This is the model I used for training the encoder and decoder, I use the FilterBank API instead of the AdaptiveEncoderDecoder in [https://github.com/mpariente/asteroid/blob/master/egs/wham/TwoStep/model.py](url). I only train the first step, see the attached logs for training. I use multisrc_neg_sisdr loss function during training. It seems that the si-sdr for custom encoder_decoder could improve but converge very very slow.
[custom_encoder_decoder.log](https://github.com/mpariente/asteroid/files/5617892/custom_encoder_decoder.log)
[train_custom_encoder_decoder.log](https://github.com/mpariente/asteroid/files/5617893/train_custom_encoder_decoder.log)

**Comment by mpariente** [comment]
Thanks for the detailed answer. 
This does look the same.

I remember that there was a padding difference, and it is now supported in `Encoder` and `Decoder`, could you try to match the ones in `AdaptiveEncoder1D` and `AdaptiveDecoder1D` please?

**Comment by xixihahaggg** [comment]
I see what happened here, I think the main reason for mismatching is the bias term. It seems that current make_enc_dec does not contain a bias term, that makes the difference. Will the filterbank API support bias in future version?

**Comment by mpariente** [comment]
That's possible. 
We could either have a `bias` flag in the `init` of the filterbank. 

Or, we could have a wrapper class `Biased` that wraps `Encoder` and `Decoder` when needed. 
How would you like the interface to be?

**Comment by xixihahaggg** [comment]
I think the first option looks simpler.

Sure, I will check the how it affect the performance.

I will close this issue and give you the results once I finish.
